binary-decode.go

```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 
 		// Read "StructTerm".
 		// NOTE: In the future, we'll need to break out of a loop
-		// when encoutering an StructTerm typ3 byte.
-		var typ = Typ3(0x00)
+		// when encountering an StructTerm typ3 byte.
+
+		var typ Typ3
+		// we have read all, or do we? consume the rest
+		for len(bz) > 2 && bz[0]&0xF8 != 0 && Typ3(bz[0]) != Typ3_StructTerm {
+			fnum, typ3, _n, _ := decodeFieldNumberAndTyp3(bz)
+			if slide(&bz, &n, _n) && err != nil {
+				return
+			}
+			if fnum > uint32(len(info.Fields)) {
+				// TODO read content but ignore
+				_, _n, err2 := consumeAny(typ3, bz)
+				if slide(&bz, &n, _n) && err2 != nil {
+					return
+				}
+			}
+		}
 		typ, _n, err = decodeTyp3(bz)
 		if slide(&bz, &n, _n) && err != nil {
 			return
 		}
 		if typ != Typ3_StructTerm {
-			err = errors.New(fmt.Sprintf("Expected StructTerm typ3 byte, got %X", typ))
+			err = errors.New(fmt.Sprintf("expected StructTerm typ3 byte, got %v", typ))
+			return
+		}
+	}
+	return
+}
+
+// read everything without doing anything with it
+func consumeAny(typ3 Typ3, bz []byte) (term bool, n int, err error) {
+	var _n int
+	switch typ3 {
+	case Typ3_Varint:
+		_, _n, err = DecodeVarint(bz)
+	case Typ3_8Byte:
+		_, _n, err = DecodeInt64(bz)
+	case Typ3_ByteLength:
+		_, _n, err = DecodeUvarint(bz)
+	case Typ3_Struct:
+		n, err = consumeStruct(bz)
+	case Typ3_StructTerm:
+		// don't slide, we'll read this later
+		term = true
+	case Typ3_4Byte:
+		_, _n, err = DecodeInt32(bz)
+	case Typ3_List:
+		_n, err = consumeList(bz)
+	case Typ3_Interface:
+		_n, err = consumeInterface(bz)
+	default:
+		panic("should not happen")
+
+	}
+
+	slide(&bz, &n, _n)
+	return
+}
+
+func consumeStruct(bz []byte) (n int, err error) {
+	var _n, typ = int(0), Typ3(0x00)
+FOR_LOOP:
+	for {
+		typ, _n, err = consumeFieldKey(bz)
+		if slide(&bz, &n, _n) && err != nil {
+			return
+		}
+		var term bool
+		term, _n, err = consumeAny(typ, bz)
+		if slide(&bz, &n, _n) && err != nil {
 			return
 		}
+		if term {
+			break FOR_LOOP
```

Comment:
line jumps makes for spaghetti code, avoid at all costs - in this situation you can just use break and then delete the like with `FOR_LOOP:`

```golang
if term {
	break
}
```

Reply:
done (replace `break` with `return` though, b/c that's what happens after the for loop)